packages/lib/src/verify.ts

```diff
@@ -1,199 +1,180 @@
+  } else {
+    const { pubKeys, hashFn } = merkleRootArgs!;
+    const edwardsPubKeys = pubKeys.map((pubKey) => pubKey.toEdwards());
+    resolvedMerkleRoot = await computeMerkleRoot(edwardsPubKeys, hashFn);
   }
-  console.timeEnd("Merkle Root Verification");
-
-  console.time("T and U Verification");
-  const [Tx, Ty, Ux, Uy] = publicSignals.slice(2, 6).map(BigInt);
-  const T = new EdwardsPoint(Tx, Ty);
-  const U = new EdwardsPoint(Ux, Uy);
-  const { R, msgHash } = proof;
-  const { T: computedT, U: computedU } = await recoverTUFromProof(R, msgHash);
-  if (!computedT.equals(T) || !computedU.equals(U)) {
-    return false;
+  if (resolvedMerkleRoot !== publicSignals.merkleRoot) {
+    return { verified: false };
   }
-  console.timeEnd("T and U Verification");
+  enableTiming && console.timeEnd(`Merkle Root Verification: ${timingUuid}`);
 
-  console.time("Nullifier Verification");
-  const proofNullifierRandomness = BigInt(publicSignals[6]);
-  if (proofNullifierRandomness !== nullifierRandomness) {
-    return false;
+  enableTiming && console.time(`T and U Verification: ${timingUuid}`);
+  const { T, U } = computeTUFromR(proof.R, proof.msgHash);
+  if (!T.equals(publicSignals.T) || !U.equals(publicSignals.U)) {
+    return { verified: false };
   }
-  console.timeEnd("Nullifier Verification");
+  enableTiming && console.timeEnd(`T and U Verification: ${timingUuid}`);
 
-  console.time("Fetching Verification Key");
+  enableTiming && console.time(`Nullifier Verification: ${timingUuid}`);
+  if (sigNullifierRandomness !== publicSignals.sigNullifierRandomness) {
+    return { verified: false };
+  }
+  if (
+    usedSigNullifiers &&
+    usedSigNullifiers.includes(publicSignals.sigNullifier)
+  ) {
+    return { verified: false };
+  }
+  enableTiming && console.timeEnd(`Nullifier Verification: ${timingUuid}`);
+
+  enableTiming && console.time(`Fetching Verification Key: ${timingUuid}`);
   const vKey = isNode()
     ? await getVerificationKeyFromFile(pathToCircuits!)
     : await getVerificationKeyFromUrl();
-  console.timeEnd("Fetching Verification Key");
+  enableTiming && console.timeEnd(`Fetching Verification Key: ${timingUuid}`);
 
-  console.time("ZK Proof Verification");
+  enableTiming && console.time(`ZK Proof Verification: ${timingUuid}`);
   const verified = await verifyMembershipZKP(vKey, proof.zkp);
-  console.timeEnd("ZK Proof Verification");
-
-  // snarkjs will not terminate this object automatically
-  // We should do so after all proving/verification is finished for caching purposes
-  // See: https://github.com/iden3/snarkjs/issues/152
-  // @ts-ignore
-  await globalThis.curve_bn128.terminate();
-  console.timeEnd("Membership Proof Verification");
-
-  return verified;
+  if (!verified) {
+    return { verified: false };
+  }
+  enableTiming && console.timeEnd(`ZK Proof Verification: ${timingUuid}`);
+  enableTiming &&
+    console.timeEnd(`Membership Proof Verification: ${timingUuid}`);
+
+  return {
```

Comment:
We should also likely be returning the pubKeyNullifier and the pubKeyNullifierRandomnessHash?

Reply:
For any proofs we're making that require them

Reply:
Think this makes more sense to do out of this library, since it is inherently dependent on a specific type of proof being made (batch proof for distinct pubkeys) - besides, this check can be done using the proof public signals, see: https://github.com/project-vinyl/jubmoji.quest/pull/19